src/tracker.hpp

```diff
@@ -23,71 +23,300 @@ See the Apache License Version 2.0 for the specific language governing permissio
+    /**
+     * @brief Identifies the specific object being actioned.
+     * 
+     * E.g., ID of the video being played, or the SKU or the product added-to-basket.
+     */
     string *label;
+  
+    /**
+     * @brief Describes the object or the action performed on it.
+     * 
+     * This might be the quantity of an item added to basket
+     */
     string *property;
+
+    /**
+     * @brief Quantifies or further describes the user action.
+     * 
+     * This might be the price of an item added-to-basket, or the starting time of the video where play was just pressed.
+     */
     double *value;
+
+    /**
+     * @brief Unix timestamp (in ms) when the event was created. Assigned automatically.
+     */
     unsigned long long timestamp;
+
+    /**
+     * @brief ID of the event (UUID v4) that is assigned automatically.
+     */
     string event_id;
+
+    /**
+     * @brief Optional, user-defined Unix timestamp (in ms) for the event to override the automatically assigned one.
+     */
     unsigned long long *true_timestamp;
+
+    /**
+     * @brief Context entities added to the event.
+     * 
+     */
     vector<SelfDescribingJson> contexts;
+
+    /**
+     * @brief Construct a new Structured Event object
+     * 
+     * @param category Name for the group of objects you want to track e.g. "media", "ecomm".
+     * @param action Defines the type of user interaction for the web object.
+     */
     StructuredEvent(string category, string action);
   };
 
+  /**
+   * @brief Event to track custom information that does not fit into the out-of-the box events.
+   * 
+   * Self-describing events are a [data structure based on JSON Schemas](https://docs.snowplowanalytics.com/docs/understanding-tracking-design/understanding-schemas-and-validation/)
+   * and can have arbitrarily many fields.
+   * To define your own custom self-describing event, you must create a JSON schema for that
+   * event and upload it to an [Iglu Schema Repository](https://github.com/snowplow/iglu) using
+   * [igluctl](https://docs.snowplowanalytics.com/docs/open-source-components-and-applications/iglu/)
+   * (or if a Snowplow BDP customer, you can use the
+   * [Snowplow BDP Console UI](https://docs.snowplowanalytics.com/docs/understanding-tracking-design/managing-data-structures/)
+   * or [Data Structures API](https://docs.snowplowanalytics.com/docs/understanding-tracking-design/managing-data-structures-via-the-api-2/)).
+   * Snowplow uses the schema to validate that the JSON containing the event properties is well-formed.
+   */
   class SelfDescribingEvent {
   public:
+    /**
+     * @brief Main properties of the self-describing event including it's schema and body
+     */
     SelfDescribingJson event; // required
+
+    /**
+     * @brief Unix timestamp (in ms) when the event was created. Assigned automatically.
+     */
     unsigned long long timestamp;
+
+    /**
+     * @brief ID of the event (UUID v4) that is assigned automatically.
+     */
```

Comment:
As comment above about deprecating event_id and timestamp.

I'm wondering if there is a technical reason that prevented these event event classes to extend a unique event base class with event_id and timestamp in the public interface.

Reply:
Yeah, I was also wondering why there is no Event base class or interface. The APIs are definitely worth cleaning up in the next version!